tests/test_cases/shutdown_fork.c

```diff
@@ -0,0 +1,130 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+#include <time.h>
+
+int main(int argc, char *argv[]) {
+	int server_fd, new_socket, client_fd; 
+	int opt = 1, opt2 = 1;
+	int dummy = 0;
+	clock_t t;
+	struct sockaddr_in address, serv_addr;
+	int addrlen = sizeof(address);
+	char buffer[32];
+
+	int parent_to_child[2];
+	
+	if (pipe(parent_to_child) < 0)
+	{
+		printf("Failed to initialize parent to child pipe");
+		fflush(stdout);
+		exit(EXIT_FAILURE);
+	}
+
+	if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+		printf("server socket failed\n");
+		fflush(stdout);
+		exit(EXIT_FAILURE);
+	}
+
+	if (fork() == 0) {
+		close(parent_to_child[1]);
+
+		if ((client_fd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+			printf("client socket failed\n");
+			fflush(stdout);
+			exit(EXIT_FAILURE);
+		}
+
+		if (setsockopt(client_fd, SOL_SOCKET, SO_REUSEADDR, &opt2, sizeof(opt2))) {
+			printf("client setsockopt failed\n");
+			fflush(stdout);
+			exit(EXIT_FAILURE);
+		}
+
+		serv_addr.sin_family = AF_INET;
+		serv_addr.sin_port = htons(5000);
+
+		if (inet_pton(AF_INET, "127.0.0.1", &serv_addr.sin_addr) <= 0) {
+			printf("inet_pton failed\n");
+			fflush(stdout);
+			exit(EXIT_FAILURE);
+		}
+
+		if (connect(client_fd, (struct sockaddr*)&serv_addr, sizeof(serv_addr)) < 0) {
+			printf("client connection failed\n");
+			fflush(stdout);
+			exit(EXIT_FAILURE);
+		}
+
+		// Blocking read pipe that continues when parent is about to read socket
+		read(parent_to_child[0], &dummy, sizeof(dummy));
+		
+		// Wait for parent to start read socket
+		sleep(3);
```

Comment:
Does this sleep need to be so long, or is it possible to sync without a sleep here?

Reply:
Essentially we need to let it shutdown after the parent process starts receiving or reading from the socket. I don't a way for us to just sync them up.

3 seconds might be too longer but since it's a test case I don't think it matters too much. But we can definitely make it 1 second.